<p>
<font size='5' face='Georgia, Arial'>IIC-2233 Apunte Programación Avanzada</font><br>
<font size='1'> Material creado por Equipo Docente IIC2233. Modificado en 2021-1 al 2023-1 por Equipo Docente IIC2233</font>
</p>

# Tabla de contenidos

1. [API (*Application Programming Interface*)](#API-(Application-Programming-Interface))
2. [HTTP (***Hypertext Transfer Protocol***)](#HTTP-(Hypertext-Transfer-Protocol))
3. [*Client-side script*](#Client-side-script)
    1. [Ejemplo consumiendo una API (método get)](#Ejemplo-consumiendo-una-API-(método-get))
    2. [Retorno de una API](#Retorno-de-una-API)
    3. [Uso de `post`](#Uso-de-post)
    4. [Autenticación en _headers_](#Autenticación-en-headers)
    5. [Caso aplicado: API de GITHUB](#Caso-aplicado:-API-de-GITHUB)
4. [*Server-side App*](#Server-side-App)

Durante el capítulo de *networking* aprendimos el uso de *sockets* y algunos protocolos para establecer la comunicación a través de una arquitectura cliente-servidor. En este capítulo, revisaremos la comunicación entre dos dispositivos mediante la **web**.

***Web services*** es el conjunto de aplicaciones cliente-servidor que se comunican a través de la web mediante un protocolo diseñado para ello. Podemos ver este tipo de servicios como aplicaciones que pueden ser accedidas por otras aplicaciones a través de una red de computadores (internet).

Por ejemplo, cuando nuestro navegador (cliente) consume un sitio web (servidor), por cada **llamada al servidor**, una aplicación escrita en algún lenguaje de programación **envía una respuesta** en [HTML](https://es.wikipedia.org/wiki/HTML) (el lenguaje que se utiliza para definir la estructura de un sitio web) para que nuestro navegador la despliegue. Los _web services_ funcionan de forma similar, donde la salida está dirigida a una **aplicación que consume** esta información. Para que la comunicación sea posible, el formato de los mensajes debe ser conocido por ambas partes, para que la información pueda ser interpretada correctamente.


## API (*Application Programming Interface*)

En general se conoce como **API** al conjunto de funciones que son expuestas por un servicio para ser utilizadas por otros programas. Podemos ver al servicio como una clase, y a la API como el conjunto de métodos de esa clase. El servicio puede ser un *web service* o cualquier paquete que exponga una interfaz, por ejemplo una librería de Python. Sin embargo, en este capítulo cuando hablemos de API nos estaremos refiriendo a los métodos expuestos por un servicio web (un *web service*).

Una gran parte de los servicios actuales exponen una API, y permiten que otras aplicaciones se conectan a ellas. De esta manera podemos construir aplicaciones que utilizan servicios que se encuentren en computadores remotos, e interactuar con ellos.

En una red de computadores, cada página web de internet es almacenada en un computador remoto que
ejecuta un proceso servidor. Un servidor remoto es simplemente un programa que escucha *requests*
(solicitudes) y envía *responses* (respuestas) de acuerdo a un protocolo.


Tú puedes utilizar tu propio computador para servir un sitio web. De hecho, los desarrolladores de *software* usan sus propios computadores como servidores locales al crear sitios web antes de publicarlos al mundo.

Cuando escribes https://www.facebook.com en tu navegador, este envía una *request* a un servidor remoto de Facebook. Una vez que tu navegador recibe la respuesta del servidor, este la interpreta y despliega una página para ti.

Para el navegador (cliente), el servidor de Facebook es una API. Esto significa que cada vez que tú visitas una página en la web, tú interactúas con alguna API en un servidor remoto. Una API no es lo mismo que un servidor remoto, pero es la parte de este que recibe las *requests* y envía *responses* (respuestas).

Si cada objeto tecnológico de tu casa expusiera una API, podrías controlarla completamente desde tu celular o desde cualquier programa en Python u otro lenguaje (ver más en [internet de las cosas](https://es.wikipedia.org/wiki/Internet_de_las_cosas)).


## HTTP (***Hypertext Transfer Protocol***)

Gran parte de las arquitecturas de *web services* se basan en el uso del protocolo **HTTP**. Este protocolo de aplicación está encargado de proporcionar una capa para realización de transacciones y así permitir la comunicación entre clientes y servidores. HTTP trabaja como un protocolo ***request-response*** en donde el cliente hace una solicitud (*request*) y el servidor responde con la información solicitada (*response*).

HTTP es un protocolo en el que el servidor no guarda ninguna información de las conexiones. Por ejemplo, al acceder a métodos de un servicio web que requiere identificación del cliente, este deberá en cada consulta enviar **algo** que acredite su identidad.

El funcionamiento de este protocolo se basa en la definición de métodos o verbos que indican la acción a desarrollar por un determinado recurso. Los recursos pueden ser datos existentes en el servidor (por ejemplo, archivos) o bien una salida generada dinámicamente. La versión HTTP/1.1 incluye **cinco** métodos:

- `GET`: recupera una representación de un recurso sin cambiar nada en el servidor.
- `POST`: crea un recurso.
- `PATCH`: aplica modificaciones parciales a un recurso.
- `PUT`: reemplaza completamente un recurso existente.
- `DELETE`: elimina un recurso.

HTTP incluye también un conjunto de códigos de estado mediante los cuales se entrega información al cliente sobre el resultado de su petición. Algunos códigos comunes de respuesta son:

- `200` : OK. Solicitud exitosa.
- `403` : Prohibido. La petición es aceptada, pero el servidor rechaza responderla.
- `404` : No encontrado. El recurso solicitado no ha sido encontrado.
- `500` : Error interno del servidor.

Para más detalle de los códigos pueden revisar el siguiente [enlace]( http://www.w3schools.com/tags/ref_httpmessages.asp).

La siguiente figura muestra un ejemplo con la estructura de los mensaje HTTP para la *request* del cliente y para la *response* desde el servidor.

![](imgs/http_message.png)


## *Client-side script*

En esta sección veremos, desde el punto de vista del cliente, cómo efectuar *requests* a un servidor que mantiene un servicio web. En Python, la librería `requests` nos permite interactuar con servicios disponibles en algún *web service*. La librería, además, integra los métodos para serialización en JSON.

Para instalar la librería `requests`, en cualquier terminal debes correr el comando `pip3 install requests`. Otra opción, es ejecutar la siguiente celda, que instalará la librería en el mismo entorno en que estés corriendo este jupyter (de todas maneras recomendamos instalarlo desde la consola):


In [3]:
import requests

Para generar una petición mediante `GET` usamos el método `get(url)` que recibe por argumento el llamado al recurso.


In [5]:
# Esta url contiene la dirección del web service 
# y los parámetros que se requiren para la consulta
url = 'https://api.github.com/repos/IIC2233/Syllabus/issues/8'
response = requests.get(url)

In [2]:
response?

Type:        Response
String form: <Response [200]>
File:        c:\users\mativ\appdata\local\programs\python\python310\lib\site-packages\requests\models.py
Docstring:  
The :class:`Response <Response>` object, which contains a
server's response to an HTTP request.

Luego podemos usar `.status_code` para saber el código  de estado de la consulta.

In [3]:
print(f'Status: {response.status_code}')

Status: 200


In [6]:
# El output de esta respuesta particular
# puede ser transformado desde JSON a dict
print(response.json()['body'])

Hola, luego de seguir los pasos para instalar Jupyter, al ejecutar el comando jupyter notebook, se me abre una ventana en blanco que contiene solo el logo de Jupyter arriba a la izquierda, mientras que en la terminal me aparece el siguiente código:

```
[I 2023-08-10 17:12:51.873 ServerApp] Package notebook took 0.0000s to import
[I 2023-08-10 17:12:51.875 ServerApp] notebook | extension was successfully linked.
[I 2023-08-10 17:12:51.913 ServerApp] notebook | extension was successfully loaded.
[I 2023-08-10 17:12:51.914 ServerApp] Serving notebooks from local directory: /Users/oliviallanos
[I 2023-08-10 17:12:51.914 ServerApp] Jupyter Server 2.7.0 is running at:
[I 2023-08-10 17:12:51.914 ServerApp] http://localhost:8888/tree?token=680d2a29734f1f0ed1f902b9403e79da37629126061ade1b
[I 2023-08-10 17:12:51.914 ServerApp]     http://127.0.0.1:8888/tree?token=680d2a29734f1f0ed1f902b9403e79da37629126061ade1b
[I 2023-08-10 17:12:51.914 ServerApp] Use Control-C to stop this server and shut dow

## Ejemplo consumiendo una API (método `get`)

A continuación vamos a experimentar con la API de [_fakerapi_](https://fakerapi.it/api/v1/). Esta nos genera difentes tipo clase de información inventada. Por ejemplo, nombre de personas, libros, etc.

En primer lugar, todas las APIs tendrán una **URL BASE**, que consiste en un _link_ de prefijo que toda consulta debe tener. Para el caso de _fakeapi_, esta URL BASE es `"https://fakerapi.it/api/v1/"`. Por lo tanto, cualquier solicitud que hagamos, debe partir con ese prefijo.

Luego, tenemos los **_endpoint_**. Para efectos del curso, podemos considerar estos **_endpoints_** como las diferentes rutas que dispone la API para hacer consultas. Para el caso de _fakeapi_, algunos **_endpoints_** son

* `books/`: nos retornará información de libros inventados.
* `texts/`: nos retornará textos inventados.
* `person/`: nos retornará información de personas inventadas.
* `addresses/` nos retornará direcciones inventadas.

Por lo tanto, para consultar esta API, deberemos combinar la **URL BASE** con un **_endpoint**. De este modo, a URL final sería `"https://fakerapi.it/api/v1/books"`, `"https://fakerapi.it/api/v1/addresses"`, entre otros. Siempre que visitemos una API, esta nos detallará los diferentes _endpoints_ que tiene.

Vamos a probar los primeros 2 **_endpoints_** de _fakerapi_.

In [8]:
BASE = "https://fakerapi.it/api/v1/"
endpoint_1 = "books/"

solicitud = requests.get(BASE + endpoint_1)

# imprimir el status_code
print(solicitud.status_code)

200


Ahora vamos a ver el contenido de la consulta

In [9]:
data = solicitud.json()
print("Llaves: ", data.keys())

Llaves:  dict_keys(['status', 'code', 'total', 'data'])


In [10]:
data["total"]

10

In [12]:
data['data'][0]

{'id': 1,
 'title': 'I\'ll be jury,".',
 'author': 'Stevie Purdy',
 'genre': 'Animi',
 'description': "March Hare. 'Yes, please do!' but the Mouse to Alice with one finger; and the White Rabbit, jumping up and repeat something now. Tell her to speak good English); 'now I'm opening out like the right.",
 'isbn': '9789600756364',
 'image': 'http://placeimg.com/480/640/any',
 'published': '1977-10-14',
 'publisher': 'Sint Dolorum'}

In [13]:
data["data"][0]["title"]

'I\'ll be jury,".'

In [15]:
data['data'][9]

{'id': 10,
 'title': "YET,' she said.",
 'author': 'Jerry Nolan',
 'genre': 'Iusto',
 'description': "It doesn't look like one, but it just now.' 'It's the first witness,' said the last concert!' on which the words 'EAT ME' were beautifully marked in currants. 'Well, I'll eat it,' said the March.",
 'isbn': '9793820287560',
 'image': 'http://placeimg.com/480/640/any',
 'published': '2011-11-03',
 'publisher': 'Et Accusantium'}

Cómo podemos notar, nos inventó el nombre, género y descripción de un libro. Ahora vamos a hacer una consulta al endpoint de `"texts/`.

In [18]:
endpoint_2 = "texts/"

solicitud = requests.get(BASE + endpoint_2)

data = solicitud.json()
data['data'][:4]

[{'title': "King, 'and don't.",
  'author': 'Diego Bednar',
  'genre': 'Repellat',
  'content': "Five! Don't go splashing paint over me like a stalk out of its right paw round, 'lives a Hatter: and in another moment, when she went in search of her voice. Nobody moved. 'Who cares for you?' said."},
 {'title': 'VERY wide, but she.',
  'author': 'Bryce Goodwin',
  'genre': 'Non',
  'content': "Alice severely. 'What are you getting on now, my dear?' it continued, turning to the door, and knocked. 'There's no such thing!' Alice was so full of smoke from one of them attempted to explain the."},
 {'title': 'THE VOICE OF THE.',
  'author': 'Tyrel Crona',
  'genre': 'Deserunt',
  'content': "As for pulling me out of court! Suppress him! Pinch him! Off with his knuckles. It was the first to speak. 'What size do you call him Tortoise--' 'Why did they live at the March Hare said to."},
 {'title': 'Alice panted as.',
  'author': 'Katrina Toy',
  'genre': 'Sed',
  'content': 'I\'ve nothing to do." S

### Parámetros de una consulta

Cada **_endpoint_** puede tener sus parámetros para personalizar la consulta. 

Por un lado, _fakeapi_ tiene algunos parámetros que sirve para cualquier **_endpoint_**. Uno de estos es `"_quantity"`, y nos permite indicar la cantidad de elementos a inventar. Por defecto es 10, pero ahora vamos a cambiarlo a 3. 

Por otro lado, el **_endpoint_** de `texts/` tiene un parámetro específico llamado `"_characters"`. Este permite indicar la cantidad de caracteres a utilizar en el texto que inventará. Por defecto es 200, pero ahora vamos a cambiarlo a 20.

In [24]:
endpoint_2 = "texts/"

parametros = {"_quantity": 1000, "_characters": 10}
solicitud = requests.get(BASE + endpoint_2, params=parametros)

data = solicitud.json()
data['data']

[{'title': 'Dormouse turned.',
  'author': 'Kevin McDermott',
  'genre': 'Ducimus',
  'content': 'Alice.'},
 {'title': 'Luckily for Alice.',
  'author': 'Aiden Heaney',
  'genre': 'Nam',
  'content': "I don't."},
 {'title': "I don't know,' he.",
  'author': 'Jaylan Schaefer',
  'genre': 'Reprehenderit',
  'content': '.'},
 {'title': "Mock Turtle. 'Very.",
  'author': 'Natalia Wiza',
  'genre': 'Perferendis',
  'content': 'There.'},
 {'title': "I wouldn't be so.",
  'author': 'Cydney Kling',
  'genre': 'Suscipit',
  'content': 'I grow.'},
 {'title': 'Alice. One of the.',
  'author': 'Trisha Parisian',
  'genre': 'Veritatis',
  'content': 'Alice.'},
 {'title': "Puss,' she began.",
  'author': 'Kennedy Tillman',
  'genre': 'Est',
  'content': 'Eaglet.'},
 {'title': 'I used--and I.',
  'author': 'Steve Nitzsche',
  'genre': 'Saepe',
  'content': 'THEIR.'},
 {'title': "I'll come up: if.",
  'author': 'Matteo Bins',
  'genre': 'Ut',
  'content': 'I am.'},
 {'title': 'There seemed to be.',
  

### Retorno de una API

Así como esta API, existen muchas otras. No todas responden en formato JSON.


In [ ]:
# Podemos usar una API para obtener nuestra IP pública
# Notar que no estamos transformando a JSON
response = requests.get('https://api.ipify.org')
ip = response.text
print(response.status_code, ip)

In [ ]:
# Podemos ahora usar una API para obtener la latitud y 
# longitud en la que nos encontramos al momento de correr este código
response = requests.get(f'https://ipapi.co/{ip}/latlong/')
print(response.status_code, response.text)

Por temas de seguridad, muchas de las APIs públicas necesitan una llave/clave para poder utilizarlas. Para conseguir estas *keys* en general debes crearte una cuenta. De esta forma se mantiene control de la aplicación expuesta, quiénes están accediendo a ella, con qué frecuencia, etc.


In [12]:
# Además, podemos usar otra API para ver más descripciones de la IP utilizada
url = 'http://api.ipstack.com/'
# En este caso puedes usar esta `API_KEY` para probar
API_KEY = 'c657ed216cf3e05d129bd6b2ccb8589e'
# Recibe la API_KEY como parámetro

# Esto puede ser enviado de dos formas:

# 1. Agregando los parámetros en la URL:
pais = requests.get('{}/{}?access_key={}'.format(url, ip, API_KEY))

# 2. Pasando los parámetros en el método:
pais = requests.get(f'{url}/{ip}', params={'access_key': API_KEY})

In [13]:
pais.status_code

200

In [14]:
pais.headers

{'content-type': 'application/json', 'transfer-encoding': 'chunked', 'date': 'Wed, 12 Apr 2023 04:35:39 GMT', 'x-apilayer-transaction-id': 'bf8e6f83-11c6-473a-a091-3525f1c4170b', 'access-control-allow-origin': '*', 'access-control-allow-methods': 'GET, POST, HEAD, OPTIONS', 'access-control-allow-headers': '*', 'x-quota-limit': '100', 'x-quota-remaining': '89', 'x-increment-usage': '1', 'x-request-time': '0.026'}

In [15]:
# En este caso la API sí retorna un JSON
pais.json()

{'ip': '181.43.208.216',
 'type': 'ipv4',
 'continent_code': 'SA',
 'continent_name': 'South America',
 'country_code': 'CL',
 'country_name': 'Chile',
 'region_code': 'RM',
 'region_name': 'Santiago Metropolitan',
 'city': 'Santiago',
 'zip': None,
 'latitude': -33.46500015258789,
 'longitude': -70.65599822998047,
 'location': {'geoname_id': 3871336,
  'capital': 'Santiago',
  'languages': [{'code': 'es', 'name': 'Spanish', 'native': 'Español'}],
  'country_flag': 'https://assets.ipstack.com/flags/cl.svg',
  'country_flag_emoji': '🇨🇱',
  'country_flag_emoji_unicode': 'U+1F1E8 U+1F1F1',
  'calling_code': '56',
  'is_eu': False}}

### Uso de `post`

En el caso de la API con la que hemos estado haciendo estas pruebas, solo se ofrecen servicios para realizar consultas, lo que se puede llevar a cabo utilizando el método `GET` del protocolo HTTP. Sin embargo, muchas veces queremos crear recursos en nuestro servidor, como por ejemplo crear un nuevo artículo para un *blog*, y para esto debemos utilizar el método `POST` del protocolo.

La API de `JSONPlaceholder` nos permite simular el uso de una API real, sin que verdaderamente exista un servicio detrás de esta. En este caso la utilizaremos para simular la creación de un artículo para un *blog*. En la práctica no estará ocurriendo nada en el servidor, puesto que es solo una simulación, pero en la vida real uno esperaría que como respuesta a nuestra *request* se cree una entrada en la base de datos del servicio que estamos utilizando. Pueden ver más información de como usar esta API [aquí](https://jsonplaceholder.typicode.com/).

A diferencia del método `GET`, cuando utilizamos el método `POST` podemos enviar información a la API utilizando el parámetro `data`, al cual podemos pasarle un diccionario de Python con la información que queremos enviar. Para este ejemplo debemos enviar la información de un artículo noticioso que queremos crear.


In [31]:
# Extracto de un artículo proveniente del blog https://venezolanoenchile.com/2016/01/20/como-es-el-clima-en-santiago-de-chile/
cuerpo = '''
El clima de Santiago es muy extraño para los que venimos de un país tropical como Venezuela y, 
más aún, para los que vivíamos en Maracaibo, como yo.\nUn día de verano, mientras caminaba a eso 
de las 2pm hacia mi trabajo, me puse a pensar en los temas que no he tocado aún en el blog. 
En ese momento, con 32°C de temperatura y bajo el sol, decidí escribir sobre este tema.

Chile es un país que tiene muchos tipos de clima, desde el desértico hasta el frío antártico. 
Pero como yo no conozco ninguna otra ciudad de Chile que no sea Santiago, todo lo que diré a 
continuación será de la capital.'
'''

data = {
    'title': '¿Cómo es el clima en Santiago de Chile?',
    'body': cuerpo,
    'userId': 1,
}


noticia = requests.post('https://jsonplaceholder.typicode.com/posts', data=data)

In [32]:
# Vemos que obtenemos un código de que nuestro artículo fue creado
print(noticia.status_code)
print(noticia.reason)

# Esta API nos retorna un JSON con el mismo recurso creado, nótese que se le asignó un id al artículo
data = noticia.json() 
print(data.keys())

201
Created
dict_keys(['title', 'body', 'userId', 'id'])


In [33]:
print(str(data['id']) + "\n")
print(data['title'] + "\n")
print(data['body'])

101

¿Cómo es el clima en Santiago de Chile?


El clima de Santiago es muy extraño para los que venimos de un país tropical como Venezuela y, 
más aún, para los que vivíamos en Maracaibo, como yo.
Un día de verano, mientras caminaba a eso 
de las 2pm hacia mi trabajo, me puse a pensar en los temas que no he tocado aún en el blog. 
En ese momento, con 32°C de temperatura y bajo el sol, decidí escribir sobre este tema.

Chile es un país que tiene muchos tipos de clima, desde el desértico hasta el frío antártico. 
Pero como yo no conozco ninguna otra ciudad de Chile que no sea Santiago, todo lo que diré a 
continuación será de la capital.'



### Autenticación en _headers_

Aparte del método `post`, tambien tenemos `.put()`, `.patch()` o `.delete()`. Generalmente, este tipo de requests repercuten en la modificación de una base de datos, y para lograr esto, muchas veces es necesario tener una autorización previa. 

Es aquí que surge la necesidad de ocupar el parámetro `headers` que permite incluir información en la cabezara de la solicitud incuyendo, por ejemplo, un token especial de acceso que a la API le sirve para identificarte y verificar si tienes los permisos. 

Utilizar este parámetro es análogo a `data`. Solo es necesario crear un diccionario e incluirlo en la _requests_. Las llaves del diccionario puede cambiar según la API que queramos acceder.

In [34]:
my_headers = {
    "Authorization": "MI-TOKEN"
}

data = {
    'title': 'Probando',
    'body': 'Probando',
    'userId': 11,
}

requests.post('https://jsonplaceholder.typicode.com/posts', headers=my_headers, data=data)

<Response [201]>

### Caso aplicado: API de GITHUB

A continuación, te presentamos un caso aplicado para crear una _issue_ en un repositorio de Github mediante su API.  En este caso, la API de Github pide que el `data` a enviar sea un diccionario en su forma de _string_, es decir, `json.dumps(data)`.

Además, para que este caso aplicado funcione correctamante, se requiere:

1. Completar `github_repo` con el nombre de tu repositorio de github dado en el curso. Por ejemplo: `"juanito-iic2233-20XX-1"`
2. Completar `token` con un _access token_ generado en Github. Puedes generar uno en [esta página](https://github.com/settings/tokens/new). Este _token_ debe tener al menos el siguiente permiso para poder crear una issue: _Full control of private repositories_.

In [ ]:
import json  # noqa: E402

github_repo = 'juanito-iic2233-20XX-1'
token = "COMPLETAR"

body = {
    'title': "Creando una issue con la API",
    'body': "Ahora tengo el poder para hacer issues desde Python! 🎉"
}

my_headers = {
    'Authorization': 'token ' + token,
    'Accept': 'application/vnd.github.v3+json'
}

url = f"https://api.github.com/repos/IIC2233/{github_repo}/issues"

response = requests.post(url, data=json.dumps(body), headers=my_headers)
response.status_code

Si la celda anterior arrojó un `201`. Significa que el _token_ utilizado y el `github_repo` estaban correctos, así que se creó la _issue_. Ejecuta la siguiente celda para ver el _link_ a dicha issue.

In [ ]:
response.json()["html_url"]

## *Server-side App*

La misión principal del servidor es disponer el contenido para que pueda ser consultado mediante un *web service*. La aplicación que corre en el servidor es la encargada de la lógica e interacción entre cliente-servidor. La información que viaja entre un cliente y un servidor permite generar comunicación entre aplicaciones.

Una aplicación puede estar desarrollada en cualquier lenguaje de programación que permita exponer una API para ser consumida por otras aplicaciones a través de la web. Por ejemplo, podemos tener una aplicación corriendo en Java, y desde nuestro código en Python acceder a esa API.

En Python existe [WSGI](https://docs.python.org/es/3/library/wsgiref.html) para exponer APIs. Tambien existen varios *frameworks* de programación que facilitan esta misma tarea, como **Flask** y **Django**. Además, puedes montar tus aplicaciones en servicios o servidores ya disponibles en la web, provistos como Platform-as-a-Service (PaaS) o Infrastructure-as-a-Service (IaaS). Por ejemplo, puedes usar **Heroku** (PaaS), **Digital Ocean** (IaaS), o **Microsoft Azure** (PaaS) para disponer tus APIs en una red pública con alta disponibilidad.


Al igual que lo hecho en *networking*, que dejábamos corriendo un servidor que estaba recibiendo solicitudes dentro del computador, podemos hacer lo mismo con *webservices* y tener un servidor corriendo que solo recibe solicitudes dentro del mismo computador.

A modo de ejemplo, tenemos una pequeña API utilizando WSGI para levantar una aplicación que responde algunos mensajes.

Antes de empezar, debes ejecutar el archivo `servidor.py` en la terminal o en VSCode para levantar esta API. Luego, puedes ejecutar las siguientes celdas donde utilizaremos `requests` para realizar solicitudes a esta API.

In [5]:
# La URL base de la API creada
BASE_URL = "http://localhost:4444/"

# Podemos consultar a esta ruta
solicitud = requests.get(BASE_URL)
solicitud.status_code

200

In [7]:
solicitud.json()

print(solicitud.json()['message'])

Hello World


In [8]:
# ¡Vamos a despedirnos haciendo una consulta a otro endpoint de nuestra API!
respuesta = requests.get(BASE_URL + "goodbye/")
print(respuesta.status_code, respuesta.json())

200 {'message': 'Que la fuerza NO esté contigo'}


**Te invitamos a consumir diferentes APIs desde Python para poner a prueba este contenidos**. Algunas ideas de estudio pueden ser:
- Revisar los diferentes _endpoint_ que tenía [fakerapi](https://fakerapi.it/en) para inventar más información.
- Revisar la documentación de [OMDb API](http://www.omdbapi.com/#usage) para buscar películas.
- Usar la APi de [pokemon](https://pokeapi.co/docs/v2#pokemon) para buscar información de tu pokemon favorito.
